src/main/java/com/scalar/db/benchmarks/tpcc/TpccBench.java

```diff
@@ -1,262 +1,159 @@
-        try {
-          tpcc.run();
-        } catch (TransactionException e) {
-          e.printStackTrace();
-        }
-      }
-      return 0;
-    }
-
-    long durationMillis = duration * 1000L;
-    long rampUpTimeMillis = rampUpTime * 1000L;
-
-    AtomicBoolean isRunning = new AtomicBoolean(true);
-    ExecutorService executor = Executors.newFixedThreadPool(numThreads);
-    final long start = System.currentTimeMillis();
-    long from = start;
-    for (int i = 0; i < numThreads; ++i) {
-      executor.execute(
-          () -> {
-            TpccRunner tpcc = new TpccRunner(manager, config);
-            while (isRunning.get()) {
-              try {
-                long eachStart = System.currentTimeMillis();
-                tpcc.run(isRunning, errorCounter);
-                long eachEnd = System.currentTimeMillis();
-                counter.incrementAndGet();
-                if (System.currentTimeMillis() >= start + rampUpTimeMillis) {
-                  totalCounter.incrementAndGet();
-                  latencyTotal.addAndGet(eachEnd - eachStart);
-                }
-              } catch (TransactionException e) {
-                if (verbose) {
-                  e.printStackTrace();
-                }
-                errorCounter.incrementAndGet();
-              }
-            }
-          });
-    }
-
-    long end = start + rampUpTimeMillis + durationMillis;
+  @Override
+  public void executeEach() throws TransactionException {
+    TpccTransaction transaction = generateTpccTransaction();
     while (true) {
-      long to = System.currentTimeMillis();
-      if (to >= end) {
-        isRunning.set(false);
+      try {
+        transaction.execute();
+        transaction.commit();
         break;
+      } catch (CrudConflictException | CommitConflictException e) {
+        transaction.abort();
+        abortCounter.incrementAndGet();
+        Uninterruptibles.sleepUninterruptibly(tpccConfig.getBackoff(), TimeUnit.MILLISECONDS);
+      } catch (Exception e) {
+        transaction.abort();
+        throw e;
       }
-      System.out.println(((double) counter.get() * 1000 / (to - from)) + " tps");
-      counter.set(0);
-      from = System.currentTimeMillis();
-
-      Uninterruptibles.sleepUninterruptibly(1000, TimeUnit.MILLISECONDS);
     }
-    System.out.println(
-        "TPS: " + (double) totalCounter.get() * 1000 / (end - start - rampUpTimeMillis));
-    System.out.println("Average-Latency(ms): " + (double) latencyTotal.get() / totalCounter.get());
-    System.out.println("Error-Counts: " + errorCounter.get());
+  }
 
-    executor.shutdown();
-    executor.awaitTermination(10, TimeUnit.SECONDS);
+  @Override
+  public void close() {
+    setState(Json.createObjectBuilder().add("abort_count", abortCounter.toString()).build());
     manager.close();
+  }
+
+  private TpccTransaction generateTpccTransaction() throws TransactionException {
+    int x = TpccUtil.randomInt(1, 100);
```

Comment:
It looks like it uses `Random`, but maybe we should use `ThreadLocalRandom` for performance reasons.

Reply:
Fixed in 2286bc9.